src/engine/guidance/post_processing.cpp

```diff
@@ -319,41 +328,63 @@ void trimShortSegments(std::vector<RouteStep> &steps, LegGeometry &geometry)
-
+        // fixup the coordinate
         geometry.locations.erase(geometry.locations.begin());
 
-        BOOST_ASSERT(geometry.segment_offsets[1] == 1);
-        // geometry offsets have to be adjusted. Move all offsets to the front and reduce by one.
-        std::transform(geometry.segment_offsets.begin() + 1, geometry.segment_offsets.end(),
-                       geometry.segment_offsets.begin(), [](const std::size_t val)
-                       {
-                           return val - 1;
-                       });
-        geometry.segment_offsets.pop_back();
-
         // remove the initial distance value
         geometry.segment_distances.erase(geometry.segment_distances.begin());
 
-        // update initial turn direction/bearings. Due to the duplicated first coordinate, the
-        // initial bearing is invalid
-        designated_depart.maneuver = detail::stepManeuverFromGeometry(
-            TurnInstruction::NO_TURN(), WaypointType::Depart, geometry);
-
-        // finally remove the initial (now duplicated move)
-        steps.erase(steps.begin());
+        // We have to adjust the first step both for its name and the bearings
+        if (zero_length_step)
+        {
+            // move offsets to front
+            BOOST_ASSERT(geometry.segment_offsets[1] == 1);
+            // geometry offsets have to be adjusted. Move all offsets to the front and reduce by
+            // one.
+            std::transform(geometry.segment_offsets.begin() + 1, geometry.segment_offsets.end(),
+                           geometry.segment_offsets.begin(), [](const std::size_t val)
+                           {
+                               return val - 1;
+                           });
+
+            geometry.segment_offsets.pop_back();
+            const auto &current_depart = steps.front();
+            auto &designated_depart = *(steps.begin() + 1);
+
+            // FIXME this is required to be consistent with the route durations. The initial turn is
+            // not
+            // actually part of the route, though
+            designated_depart.duration += current_depart.duration;
+
+            // update initial turn direction/bearings. Due to the duplicated first coordinate, the
+            // initial bearing is invalid
+            designated_depart.maneuver = detail::stepManeuverFromGeometry(
+                TurnInstruction::NO_TURN(), WaypointType::Depart, geometry);
+
+            // finally remove the initial (now duplicated move)
+            steps.erase(steps.begin());
+        }
+        else
+        {
+            steps.front().geometry_begin = 1;
+            // reduce all offsets by 1
+            std::transform(geometry.segment_offsets.begin(), geometry.segment_offsets.end(),
+                           geometry.segment_offsets.begin(), [](const std::size_t val)
+                           {
+                               return val - 1;
+                           });
```

Comment:
these two are not exactly identical. One transforms in place, one moves to the front in addition. Maybe clarify with an additional comment?